docs/circuits.ipynb

```diff
@@ -579,6 +579,137 @@
     "Especially useful is dropping the last moment (which are often just measurements): `circuit[:-1]`, or reversing a circuit: `circuit[::-1]`.\n"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "e31b4ff57134"
+   },
+   "source": [
+    "### Nesting circuits with CircuitOperation\n",
+    "\n",
+    "Circuits can be nested inside one another with `CircuitOperation`. This is useful for concisely defining large, repetitive circuits, as the repeated section can be defined once and then be reused elsewhere. Circuits that need to be serialized especially benefit from this, as loops and functions used in the Python construction of a circuit are otherwise not captured in serialization.\n",
+    "\n",
+    "The subcircuit must first be \"frozen\" to indicate that no further changes will be made to it."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "b3772732f57b"
+   },
+   "outputs": [],
+   "source": [
+    "subcircuit = cirq.Circuit(H(q1), CZ(q0, q1), CZ(q2, q1), H(q1))\n",
+    "subcircuit_op = cirq.CircuitOperation(subcircuit.freeze())\n",
+    "circuit = cirq.Circuit(H(q0), H(q2), subcircuit_op)\n",
+    "print(circuit)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "3557398496c2"
+   },
+   "source": [
+    "Frozen circuits can also be constructed directly, for convenience."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "3d80776e09fe"
+   },
+   "outputs": [],
+   "source": [
+    "circuit = cirq.Circuit(\n",
+    "    cirq.CircuitOperation(\n",
+    "        cirq.FrozenCircuit(H(q1), CZ(q0, q1), CZ(q2, q1), H(q1))\n",
+    "    )\n",
+    ")\n",
+    "print(circuit)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "e2c83225f05a"
+   },
+   "source": [
+    "A `CircuitOperation` is sort of like a function: by default, it will behave like the circuit it contains, but we can also pass arguments to it that alter the qubits it operates on, the number of times it repeats, and other properties. `CircuitOperation`s can also be referenced multiple times within the same \"outer\" circuit for conciseness."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "62d50a44bdb3"
+   },
+   "outputs": [],
+   "source": [
+    "subcircuit_op = cirq.CircuitOperation(cirq.FrozenCircuit(CZ(q0, q1)))\n",
+    "repeated_subcircuit_op = subcircuit_op.repeat(2)\n",
+    "moved_subcircuit_op = subcircuit_op.with_qubit_mapping({q0: q2})\n",
+    "circuit = cirq.Circuit(\n",
```

Comment:
It might be useful to have some line comments to further explain this example.

Reply:
Done.